internal/pipeline/runtime/container.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"fmt"
+	"io"
+	"os/user"
+)
+
+// ContainerNetworkName is a type for network name used in container
+type containerNetworkName string
+
+const (
+	networkNameNone containerNetworkName = "none"
+	networkNameHost containerNetworkName = "host"
+)
+
+type containerFunctionPermission struct {
+	AllowNetwork  bool
+	AsCurrentUser bool
+}
+
+// ContainerRunner implements a KRMFn which run a containerized
+// KRM function
+type ContainerRunner struct {
+	// Image is the container image to run
+	Image string
+
+	Exec ExecRunner
+
+	containerFunctionPermission
+}
+
+// Run implements KRMFn. It will run the container function with
+// stdin from r and write the output to w
+func (f *ContainerRunner) Run(r io.Reader, w io.Writer) error {
+	err := f.setupExec()
+	if err != nil {
+		return fmt.Errorf("error when setup exec: %w", err)
+	}
+	return f.Exec.Run(r, w)
+}
+
+func (f *ContainerRunner) setupExec() error {
+	// don't init 2x
+	if f.Exec.Path != "" {
+		return nil
+	}
+
+	path, args, err := f.getCommand()
+	if err != nil {
+		return fmt.Errorf("error when get command and args: %w", err)
+	}
+	f.Exec.Path = path
+	f.Exec.Args = args
+	return nil
+}
+
+// getCommand returns the command + args to run to spawn the container
+func (f *ContainerRunner) getCommand() (string, []string, error) {
```

Comment:
may be have this function return `exec.Cmd`

Reply:
I think return an ExecFn may be better? Or just merge this function with `setupExec` and directly setup the `Exec` member. `exec.Cmd` should be encapsulated by `ExecFn`.

Reply:
I shared my view in an earlier comment. If we decide to go that route, this can be revisited.